Introduce GitLab Driver

GromNaN opened the pull request:
Following the work of @henrikbjorn on https://github.com/composer/composer/pull/3369 to build a GitLab repository driver. Fixes #3241

I've tested with a simple project like the following, and on a Satis instance with 150 large repositories. The performance impact is rather huge! 

```json
# composer.json
{
    "repositories": [
        { "type": "gitlab", "url": "http://gitlab.example.com/mirror/symfony_http_foundation" }
    ],
    "config": {
        "gitlab-domains": ["gitlab.example.com"]
    },
    "require": {
        "symfony/http-foundation": "2.6.*"
    }
}
```

* The config `gitlab-domains` is required for custom GitLab servers. It is set to [**gitlab.com**](https://about.gitlab.com/gitlab-com/) by default for people using this host.
* The repository type "gitlab" is optional here. If the the type is `vcs`, the repository is properly detected based on the gitlab domains configuration.
* The authentication credential are asked interactively and the oauth token is stored in `auth.json`. This token is required to scan the repository and to download archives.

```json
# ~/.composer/auth.json
{
    "gitlab-oauth": {
        "gitlab.example.com": "597d83f59332a9de4795b2..."
    }
}
```

* It makes a real boost in terms of repository scanning as it makes only 3 API calls when everything is already in the cache : one for the project infos, and 2 others to list branches and tags. All the `composer.json` files are stored in the cache.

<del>**Side note:** I've used [Prophecy](https://github.com/phpspec/prophecy) for unit tests, [it comes out-of-the box with PHPUnit 4.5](http://thephp.cc/news/2015/02/phpunit-4-5-and-prophecy). But that imposes to upgrade the required version of `phpunit/phpunit`. https://github.com/composer/composer/pull/3766</dev>

stof commented on doc/04-schema.md:
```diff
@@ -798,6 +798,9 @@ The following options are supported:
   tokens created to access the github API will have a date instead of the
   machine hostname.
+* **gitlab-domains:** Defaults to `["gitlab.com"]`. A list of domains to use pointing
+  to GitLab servers. This is used if you host your own GitLab and use `gitlab`
```
missing word: ``use the gitlab repository type``

stof commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,278 @@
+    protected $project;
+    protected $commits = array();
+    protected $tags;
+    protected $branches;
```
I suggest using private visibility

stof commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,278 @@
+        $this->scheme       = $match[1];
+        $this->originUrl    = $match[2];
+        $this->owner        = $match[3];
+        $this->repository   = $match[4];
```
the ``=`` signs should not be aligned here according to the Symfony CS.

stof commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,278 @@
+     */
+    public function getApiUrl()
+    {
+        // this needs to be https, but our install is running http
```
this comment looks weird here

GromNaN replied:
Indeed. I've removed this relic and fixed CS.

henrikbjorn commented:
@GromNaN this is so awesome, thanks for picking it up :)!

stof commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,294 @@
+        // Convert the root identifier to a cachable commit id
+        if (!preg_match('{[a-f0-9]{40}}i', $identifier)) {
+            foreach ($this->getBranches() as $ref => $id) {
+                if ($ref === $identifier) {
```
doing a loop over branches to compare keys is really weird. Use a key lookup instead (which is ``O(1)`` while your loop is ``O(n)`` btw)

GromNaN replied:
Thanks. Fixed.

stof commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,294 @@
+            $references[$datum['name']] = $datum['commit']['id'];
+
+            // Keep the last commit date of a reference to avoid
+            // unnecessary API call when retreiving the composer file.
```
typo: ``retrieving``

GromNaN commented:
Fixed tests.
Can anyone test it against its own GitLab repositories ?

GromNaN commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,292 @@
+    /**
+     * {@inheritDoc}
+     */
+    public function hasComposerFile($identifier)
```
I might remove this function as it's basically the same as its parent.

GromNaN commented:
I've rebased the branch and squashed some commits.

To prove that it  works as expected, I've created a satis repository that scans a [symfony mirror on GitLab.com](https://gitlab.com/GromNaN/symfony). The commit is here : https://github.com/GromNaN/satis/commit/0b450df3335dd7f85501fba4f0a506c6156eafbe.

During the first build, it asks for the GitLab Token:

```
$ php bin/satis build config.json web
Scanning packages
Enter your GitLab private token to access API (gitlab.com):
      Private-Token: ************
wrote packages json web/include/all$2ae63618c7d4de5c5a3fe0063dc4fcaa831a575f.json
Writing packages.json
Writing web view
```

Is there anything that I can do to make this PR accepted ?

wedgwood commented:
awesome +1

achretien commented:
tested successfully on our company private Gitlab server (> 200 repos)

vinkla commented:
This would be awesome. Plus one on GitLab support!

roshangautam commented:
This looks promising. Can anyone here show me how to configure satis for dist archive pulling from gitlab api? Should i put the following in prefix-url in archive section ?

http://somegitlabserver.com/api/v3/projects/:id/repository/archive?private_token

This is what i want. When composer install or update is run it should ask for gitlab username and password and if authenticated the private token of that user should be used to download archive. Doest that make sense ?

GromNaN commented:
You must configure your gitlab domain like this:

```json
# composer.json
{
    "config": {
        "gitlab-domains": ["gitlab.com", "somegitlabserver.com"]
    },
}
```

Then composer will ask for your private token directly.

roshangautam commented:
+1. This needs to be merged.

roshangautam commented:
@GromNaN Check this out: The api to fetch project using NAMESPACE/PROJECT_NAME is not working in gitlab and hence creating problems.

https://github.com/gitlabhq/gitlabhq/issues/8290

GromNaN commented:
Thanks @roshangautam for the link to the issue. This is not a real bug of GitLab, but more due to the way people have configured their webserver.

I had a similar issue with Apache as reverse proxy. I fixed it like this:
https://gitlab.com/gitlab-org/gitlab-recipes/commit/73abde7501b6cd128ce22ea9f82dc19fdc867c80

roshangautam commented:
@GromNaN I added the AllowEncodedSlashes NoDecode to my vhost configuration file but now am getting project not found error. Any idea ?

brunogaspar commented:
+1 for Gitlab support!

GromNaN commented:
The authentication can be made with OAuth; but that requires a bugfix in GitLab. https://github.com/GromNaN/composer/pull/1

achretien commented:
Is there any update about this request ?

ikappas commented:
Hi all, I have also issued a merge request on gitlab ce for adding a Project Composer Service. More details and/or discussion https://github.com/gitlabhq/gitlabhq/pull/9301

tritumRz commented:
+1
I tested it with my own gitlab instance and it works great.

wedgwood commented:
come on~

Seldaek commented:
@GromNaN sorry for the delay on this, but I had a look and it looks fine to me, but before merging I'd like to clarify whether https://github.com/GromNaN/composer/pull/1 is ready or not, because it changes the gitlab-tokens into gitlab-oauth and seems to change quite a bit of how it works with regard to auth. So I'd rather wait for that if you think it's a better way to do auth than the current model. I am not very familiar with gitlab so not sure.

mishak87 commented on src/Composer/Util/RemoteFilesystem.php:
```diff
@@ -410,6 +421,8 @@ protected function getOptionsForUrl($originUrl, $additionalOptions)
             $auth = $this->io->getAuthentication($originUrl);
             if ('github.com' === $originUrl && 'x-oauth-basic' === $auth['password']) {
                 $options['github-token'] = $auth['username'];
+            } elseif ($auth['password'] === 'gitlab-private-token') {
```
Shouldn't be there origin check too?
```php
if (in_array($this->originUrl, $this->config->get('gitlab-domains'), true) && 'gitlab-private-token' === $auth['password']) {
```

mishak87 commented:
:+1: This pull will solve some of my issues, if you need help or more testing just ping me.

GromNaN commented:
@mishak87 Can you test https://github.com/GromNaN/composer/pull/1 ?

mishak87 commented:
@GromNaN Thanks will do!

mishak87 commented:
@GromNaN Protocol must be detected the http only is not ok. With wrong credentials it errors because of undefined variable $username. **With right credentials it works!** :heart_eyes_cat:

GromNaN commented:
Yeah ! That works and this is awesome.
I merged OAuth feature and made extensive tests (on my PHP 5.3).

@Seldaek this is ready to be merged.

GromNaN commented:
There is 2 issues related to the undefined scheme of the GitLab instance.

1. The Git URL cannot be an SSH URL as we need to detect the scheme (https://github.com/composer/composer/pull/3765/files#diff-02b1295e0c68eccffa9605b12c1015d1R72)
2. The OAuth request is using `http` ; we need to pass the scheme from `GitLabDriver` to `GitLab`. (https://github.com/composer/composer/pull/3765/files#diff-05f3e1829e608fea85d7b258821d100cR146)

mishak87 commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,367 @@
+
+    /**
+     * Extracts information from the repository url.
+     * SSH urls are not supported in order to know the HTTP sheme to use.
```
`sheme` => `scheme`

mishak87 commented:
Ad 2. could it be implemented like GitHub is with [github-protocols](https://github.com/GromNaN/composer/blob/gitlab/doc/04-schema.md#config-root-only) [git, https, http]?

GromNaN commented:
As suggested by @mishak87, I made SSH urls using HTTPS by default. But I can't fallback to HTTP.
The protocol is passed along.

This PR is finally ready.

alcohol commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,367 @@
+use Composer\Json\JsonFile;
+use Composer\Downloader\TransportException;
+use Composer\Util\RemoteFilesystem;
+use Composer\Util\GitLab;
```
PSR2: Blank line after last `use` statement.

alcohol commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,367 @@
+    private $scheme;
+    private $owner;
+    private $repository;
+
```
Why is there a blank line here suddenly?

alcohol commented on src/Composer/Util/RemoteFilesystem.php:
```diff
@@ -463,6 +487,10 @@ protected function getOptionsForUrl($originUrl, $additionalOptions)
             if ('github.com' === $originUrl && 'x-oauth-basic' === $auth['password']) {
                 $options['github-token'] = $auth['username'];
+            } elseif ($this->config && in_array($originUrl, $this->config->get('gitlab-domains'), true)) {
+                if($auth['password'] === 'oauth2') {
```
PSR2: Space between `if` and `(`

alcohol commented:
Some minor psr2 stuff, but other than that, implementation looks good at a quick glance.

glen-84 commented on src/Composer/Repository/Vcs/GitLabDriver.php:
```diff
@@ -0,0 +1,367 @@
+
+    protected function attemptCloneFallback()
+    {
+
```
Blank line.

Seldaek commented:
Thanks!

GromNaN commented:
Thanks to everyone who contributed to this feature.

wedgwood commented:
:beers:

ikappas commented:
Great job and thanks to everyone on this feature!

I 've run into a problem with this though in conjuction with my [Gitlab MR](https://github.com/gitlabhq/gitlabhq/pull/9301) and would like some assistance.

**Example composer.json:**
>Note that the `type: composer` of the gitlab server is required to be used as a composer repository.

```
{
    "repositories": [
        {
          "type": "composer",
          "url": "https://gitlab.example.com"
        }
      ],
    "config": {
        "gitlab-domains": ["gitlab.example.com"],
        "preferred-install": "dist"
    },
    "require": {
        "some/package": "1.0"
    }
}
```

**running `composer install`:**
```
Loading composer repositories with package information
Installing dependencies (including require-dev)
  - Installing some/package (1.0)
    Downloading: Connecting...
Could not fetch https://gitlab.example.com/some/package/repository/archive.zip?ref=1.0, enter your gitlab.example.com credentials to access private repos
A token will be created and stored in "/Users/[username]/.composer/auth.json", your password will never be stored
To revoke access to this token you can visit gitlab.example.com/profile/applications
Username: [ entered username ]
Password: [ entered password ]
Token successfully created
    Downloading: Connecting...
Could not fetch https://gitlab.example.com/some/package/repository/archive.zip?ref=1.0, enter your gitlab.example.com credentials to access private repos
A token will be created and stored in "/Users/[username]/.composer/auth.json", your password will never be stored
To revoke access to this token you can visit gitlab.example.com/profile/applications
Username: ...
```
results to an endless loop of asking credentials and creating Anonymous oAuth token under https://gitlab.example.com/profile/applications.

I'am using the latest composer and Gitlab CE 8.2.1 with [Composer Service](https://github.com/gitlabhq/gitlabhq/pull/9301)

alcohol commented:
Could it possibly be similar to #3651 ?

ikappas commented:
@alcohol I am not sure... After multiple debug sessions I can see that this happens from the fact that repositories of type `composer` do not load the gitlab driver which is expected and in turn fail to authenticate with the gitlab server over oAuth (expected?) ... 

The [composer service MR](https://github.com/gitlabhq/gitlabhq/pull/9301) basically turns GitLab into a native composer repository but I think it won't authenticate with oAuth...

I am still looking at it.

jdforsythe commented:
Can anyone who has this working successfully help out? When I try to fetch the repository with composer, the API endpoint returns the login page. It asks for the credentials and stores the access token properly. I tried hitting the same endpoint with Postman and I get the login page there, too. I can hit the main project page `http://localgit/api/v3/projects/group-name%2Fproject.name` and get back a JSON object (so I know the token is working), but any sub-paths to that redirect to the login page.

andkirby commented:
I really don't understand to set this up within satis.

*First try*
Where should I declare my gitlab domain? Ok, I added it in global config.json and in composer.json
```
{
  "config": {
    "gitlab-domains": [
      "gitlab.com",
      "git.example.com"
    ]
  }
}
```
I set auth.json in global
```
{
    "bitbucket-oauth": {},
    "github-oauth": {},
    "gitlab-oauth": {
        "git.example.com": "QERWQEWRQEWR"
    },
    "http-basic": {}
}
```
and got this:
```
$ vendor/bin/satis build satis.json html
Scanning packages
Failed to download onepica/opal:Invalid credentials for 'https://git.aaaaaaa.com/api/v3/projects/foo%2Fbar', aborting.
Your credentials are required to fetch private repository metadata (https://git.example.com/foo/bar.git)
A token will be created and stored in "C:/Users/a.user/.composer/auth.json", your password will never be stored
To revoke access to this token you can visit git.example.com/profile/applications
Username:
```
And I could not use my login/password. 
Here git.example.com/profile/applications I cannot create token. There is Application Id and Secret.
I tried to used as login/password and it doesn't work.

*Second try*
Removed auth.json in global.
I have satis.json:
```
{
  "name": "MySatis",
  "homepage": "http://satis.example.com/",
  "repositories": [
    {
      "url": "https://git.example.com/foo/bar.git",
      "type": "gitlab",
      "options":  {
        "http": {
          "header": [
            "PRIVATE-TOKEN: QWEQWEQWEQWE"
          ]
        }
      }
    }
  ],
  "require-all": true,
  "output-dir": "html"
}
```
Running:
```
$ vendor/bin/satis build satis.json -vvv
Loading config file C:/Users/a.user/.composer/config.json
Checked CA file D:\home\pkg\vendor\composer\ca-bundle\res\cacert.pem: valid
Executing command (D:\home\pkg): git branch --no-color --no-abbrev -v
Reading C:/Users/a.user/.composer/composer.json
Loading config file C:/Users/a.user/.composer/config.json
Loading config file C:/Users/a.user/.composer/composer.json
Reading D:\home\pkg/vendor/composer/installed.json
Reading C:/Users/a.user/.composer/vendor/composer/installed.json
Scanning packages
Downloading https://git.example.com/api/v3/projects/foo%2Fbar

Could not fetch https://git.example.com/api/v3/projects/foo%2Fbar, enter your git.example.com credentials to access private repos
A token will be created and stored in "C:/Users/a.user/.composer/auth.json", your password will never be stored
To revoke access to this token you can visit git.example.com/profile/applications
Username:
```

So, what's wrong?
Note: I didn't find header `PRIVATE-TOKEN: QWEQWEQWEQWE` in `$options` in vendor/composer/composer/src/Composer/Util/RemoteFilesystem.php:263

Installed composer: `dev-master 909dbf5`.

alcohol commented:
Pretty sure those `/api/v3/` urls are Github, so obviously they are not recognized as Gitlab. That would explain why your Gitlab driver setup is not working (it's not being used).

andkirby commented:
**UPDATE**
So, I've hardcoded the token before
{{vendor/composer/composer/src/Composer/Util/RemoteFilesystem.php:263}}
```
$options['http']['header'][] = 'PRIVATE-TOKEN: QWERASDFERAWE';
$ctx = StreamContextFactory::getContext($fileUrl, $options, array('notification' => array($this, 'callbackGet')));
```
And it work fine.
The question is: how to pass GitLab token? Why configured private token wasn't passed into options?

andkirby commented:
Also this line is weird for me: `Composer/Util/RemoteFilesystem.php:716`
```php
$headers[] = 'Authorization: Bearer '.$auth['username'];
```
Seem it should be
```
$headers[] = 'PRIVATE-TOKEN: '.$auth['username'];
```
instead for gitlab like it's said in [documentation](https://docs.gitlab.com/ce/api/):
```
curl --header "PRIVATE-TOKEN: 9koXpg98eAheJpvBs5tK" "https://gitlab.example.com/api/v3/projects"
```

alcohol commented:
From their documentation:

> OAuth 2 Tokens
>
> You can use an OAuth 2 token to authenticate with the API by passing it either in the access_token parameter or in the Authorization header.
>
> Example of using the OAuth2 token in the header:
>
```
curl --header "Authorization: Bearer OAUTH-TOKEN" https://gitlab.example.com/api/v3/projects
```

alcohol commented:
Seems you are not using an oauth token, but instead something else.

tritumRz commented:
composer.json:

        {
            "config": {
                "gitlab-domains": ["gitlab.example.com"],
            },
            "repositories": [
                {
                  "type": "vcs",
                  "url": "ssh://git@gitlab.example.com/namespace/repository.git"
                }
            ],
            "require": {
                "some/package": "1.0"
            }
        }

Now, composer ask you for your Gitlab credentials.
Then composer will create an oauth token within your gitlab instance for your account.
(You can see this here: https://gitlab.example.com/profile/applications)

After that your `~/.composer/auth.json` looks like this:

        {
            "gitlab-oauth": {
                "gitlab.example.com": "63745346835680c74290b62c8f6111c1a54d7c9f01aa3fb738380218e76ad85b"
            }
        }

andkirby commented:
Thanks for the answers. Yeah, seem it doesn't support private token.

But where this `composer.json` suppose to be in using `composer/satis`?
```
        "config": {
            "gitlab-domains": ["gitlab.example.com"],
        },
        "repositories": [
            {
              "type": "vcs",
              "url": "ssh://git@gitlab.example.com/namespace/repository.git"
            }
        ],
```
In satis I have only satis.json.

alcohol commented:
Please stop replying on this old pull request thread. Your questions/issue are not directly related to it.

houstona commented:
I am running into an issue with using the gitlab driver from within a repo of type composer. It says that there is no such download type.
